src/sage/rings/species.py

```diff
@@ -0,0 +1,1073 @@
+        self._cache = dict()
+
+    def clear_cache(self):
+        r"""
+        Clear the cache.
+
+        EXAMPLES::
+
+            sage: A = AtomicSpecies(1)
+            sage: A(SymmetricGroup(1))
+            {[()]: ((1,),)}
+            sage: A(SymmetricGroup(0))
+            {[]: ((),)}
+            sage: A._cache
+            {((0,), []): [{[]: ((),)}], ((1,), [()]): [{[()]: ((1,),)}]}
+            sage: A.clear_cache()
+            sage: A._cache
+            {}
+        """
+        self._cache = dict()
+
+    def _cache_get(self, elm):
+        r"""
+        Return the cached element, or add it
+        if it doesn't exist.
+
+        ``elm`` must implement the following methods:
+          ``_element_key`` - hashable type for dict lookup.
+          ``__eq__`` - to compare two elements.
+        """
+        key = elm._element_key()
+        if key in self._cache:
+            lookup = self._cache[key]
+            for other_elm in lookup:
+                if elm == other_elm:
+                    return other_elm
+            elm._canonicalize()
+            lookup.append(elm)
+        else:
+            elm._canonicalize()
+            self._cache[key] = [elm]
+        return elm
+
+class ConjugacyClassOfDirectlyIndecomposableSubgroups(Element):
+    def __init__(self, parent, C):
+        r"""
+        A conjugacy class of directly indecomposable subgroups.
+
+        TESTS::
+
+            sage: from sage.rings.species import ConjugacyClassesOfDirectlyIndecomposableSubgroups
+            sage: C = ConjugacyClassesOfDirectlyIndecomposableSubgroups()
+            sage: G = C(PermutationGroup([[(1,2),(3,4)],[(1,2),(5,6)]]))
+            sage: TestSuite(G).run()
+        """
+        Element.__init__(self, parent)
+        self._C = C
+
+    def __hash__(self):
+        r"""
+        Return the hash for the conjugacy class.
+
+        TESTS::
+
+            sage: from sage.rings.species import ConjugacyClassesOfDirectlyIndecomposableSubgroups
+            sage: C = ConjugacyClassesOfDirectlyIndecomposableSubgroups()
+            sage: G = PermutationGroup([[(1,2),(3,4),(5,6),(7,8,9,10)]]); G
+            Permutation Group with generators [(1,2)(3,4)(5,6)(7,8,9,10)]
+            sage: H = PermutationGroup([[(1,2,3,4),(5,6),(7,8),(9,10)]]); H
+            Permutation Group with generators [(1,2,3,4)(5,6)(7,8)(9,10)]
+            sage: hash(C(G)) == hash(C(H))
+            True
+        """
+        return hash(self._C)
+
+    def _repr_(self):
+        r"""
+        Return a string representation of ``self``.
+
+        EXAMPLES::
+
+            sage: from sage.rings.species import ConjugacyClassesOfDirectlyIndecomposableSubgroups
+            sage: C = ConjugacyClassesOfDirectlyIndecomposableSubgroups()
+            sage: G = PermutationGroup([[(1,3),(4,7)], [(2,5),(6,8)], [(1,4),(2,5),(3,7)]])
+            sage: C(G)
+            [(5,6)(7,8), (1,2)(3,4), (1,3)(2,4)(5,6)]
+        """
+        return f"{self._C.gens_small()}"
```

Comment:
Try to avoid `gens_small` unless necessary :-)

Reply:
True.